perfkitbenchmarker/providers/gcp/gcp_dpb_dataflow.py

```diff
@@ -160,4 +195,172 @@ def GetMetadata(self):
     basic_data = super(GcpDpbDataflow, self).GetMetadata()
     basic_data['dpb_dataflow_runner'] = FLAGS.dpb_dataflow_runner
     basic_data['dpb_dataflow_sdk'] = FLAGS.dpb_dataflow_sdk
+    basic_data['dpb_job_id'] = self.job_id
     return basic_data
+
+  @cached_property
+  def job_stats(self):
+    """Collect series of relevant performance and cost stats."""
+    stats = {}
+    stats['total_vcpu_time'] = self.GetMetricValue('TotalVcpuTime')/3600         # vCPU-hr
+    stats['total_mem_usage'] = self.GetMetricValue('TotalMemoryUsage')/1024/3600 # GB-hr
+    stats['total_pd_usage'] = self.GetMetricValue('TotalPdUsage')/3600           # GB-hr
+    # TODO(rarsan): retrieve BillableShuffleDataProcessed
+    # and/or BillableStreamingDataProcessed when applicable
+    return stats
+
+  def CalculateCost(self, type=DATAFLOW_TYPE_BATCH):
+    if type not in (DATAFLOW_TYPE_BATCH, DATAFLOW_TYPE_STREAMING):
+      raise ValueError(f'Invalid type provided to CalculateCost(): {type}')
+
+    total_vcpu_time = self.job_stats['total_vcpu_time']
+    total_mem_usage = self.job_stats['total_mem_usage']
+    total_pd_usage = self.job_stats['total_pd_usage']
+
+    cost = 0
+    if type == DATAFLOW_TYPE_BATCH:
+      cost +=  total_vcpu_time * VCPU_PER_HR_BATCH
+      cost +=  total_mem_usage * MEM_PER_GB_HR_BATCH
+    else:
+      cost += total_vcpu_time * VCPU_PER_HR_STREAMING
+      cost += total_mem_usage * MEM_PER_GB_HR_STREAMING
+    
+    cost += total_pd_usage * PD_PER_GB_HR
+    # TODO(rarsan): Add cost related to per-GB data processed by Dataflow Shuffle
+    # (for batch) or Streaming Engine (for streaming) when applicable
+    return cost
+
+  def _PullJobMetrics(self, force_refresh=False):
+    """Retrieve and cache all job metrics from Dataflow API"""
+    # Skip if job metrics is already populated unless force_refresh is True
+    if self.job_metrics is not None and not force_refresh:
+      return
+    # Raise exception if job id not available
+    if self.job_id is None:
```

Comment:
Let's better raise this error in the SubmitJob method, where you parse the ID (to make it fail fast). Also I wouldn't use ValueError here, it implies that the error is raised because a bad argument was passed (which I don't think it's necessarily the case), a generic Exception would do fine here.

Reply:
As discussed, we'll leave as is: one can submit the job and let it be acceptable not to have a job id after that, but if they try to get metrics then they get the error.

Reply:
As per our internal discussion we'll leave this alone.